File: Algorithm/Leetcode/Code_Materials/python_source/charpter1_linklist/level2/topic2_2判断回文/is_plalindrome.py
# coding=utf-8
from python_source.list_node import ListNode, init_list


class IsPalindrome:
    def isPalindrome1(self, head):
        cur, length = head, 0
        result = []
        # 遍历链表，把所有节点上的元素都存储到列表 result 中
        while cur is not None:
            length += 1
            result.append(cur.val)
            cur = cur.next
            # 定义2个指针，一个从头开始往后，另一个从最后开始往前
        left, right = 0, length - 1
        while left < right:
            if result[left] != result[right]:
                return False
            left += 1
            right -= 1
        return True

    # 方法2：双指针+反转
    def isPalindrome2(self, head):
        fake = ListNode(-1)
        fake.next = head
        fast = slow = fake
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next

        post = slow.next
        slow.next = None
        pre = head

        rev = None
        # 反转
        while post:
            tmp = post.next
            post.next = rev
            rev = post
            post = tmp

        part1 = pre
        part2 = rev
        while part1 and part2:
            if part1.val != part2.val:
                return False
            part1 = part1.next
            part2 = part2.next
        return True


if __name__ == '__main__':
    # 先构造题目要求的链表,简单构造即可
    nums = [1, 2, 2, 1]
    list = init_list(nums)
    isPalindrome = IsPalindrome()
    print isPalindrome.isPalindrome1(list)

File: Algorithm/Leetcode/Code_Materials/c_or_cpp/charpter1_linklist/level2/topic2_2判断回文/is_plalindrome.cpp
#include <stdlib.h>
#include <stdio.h>

#include <stdio.h>
#include "../../list_node.cpp"

bool isPalindrome(struct ListNode *head)
{
    if (head == NULL || head->next == NULL)
    {
        return true;
    }

    struct ListNode *slow = head;
    struct ListNode *fast = head;
    struct ListNode *pre = head;
    struct ListNode *prepre = NULL;

    while (fast != NULL && fast->next != NULL)
    {
        pre = slow;
        slow = slow->next;
        fast = fast->next->next;
        // 将前半部分链表反转
        pre->next = prepre;
        prepre = pre;
    }

    if (fast != NULL)
    {
        slow = slow->next;
    }

    while (pre != NULL && slow != NULL)
    {
        if (pre->val != slow->val)
        {
            return false;
        }
        pre = pre->next;
        slow = slow->next;
    }

    return true;
}

int main()
{
    struct ListNode *p = NULL;
    printf("create list: \t\n");
    // 创建链表0~9
    int a[] = {1, 2, 2, 1};
    p = initLink(a, 4);
    bool result = isPalindrome(p);
    printf("result=%d\n", result);

    return 0;
}